File: aoc2021/day13/af.rb
require 'set'

INPUT='input'

conns = []
folds = []
File.open( INPUT ) do |f|
	while true do
		row = f.gets.strip
		break if row.empty?
		conns << row.split(",").map(&:to_i)
	end
	while true do
		row = f.gets.strip rescue ''
		break if row.empty?
		axis,value = row[11..].split("=")
		folds << [axis=='x'?0:1, value.to_i]
	end
end

folds.each do |axis, value|
	p axis, value
	p conns
	conns = conns.map do |conn|

		conn[axis] = value-(conn[axis]-value).abs
		conn
	end.to_set.to_a
	p conns
end

p conns

w = conns.map{|x|x[0]}.max + 1
h = conns.map{|x|x[1]}.max + 1

puts w, h
m = [' '] * w * h
conns.each do |x,y|
	m[w*y+x]='W'
end

h.times.each do |r|
	puts m[(r*w)..((r+1)*w-1)].join('')
end

#631

File: aoc2021/day13/af.py
import collections, itertools

INFILE = 'test'
INFILE = 'input'

def read_input(fname):
	coords = []
	folds = []
	with open(fname) as fp:
		while True:
			row = fp.readline().strip()
			if row == "":
				break
			row = row.split(",")
			coords.append(tuple(map(int, row)))
		while True:
			row = fp.readline().strip()
			if row == "":
				break
			axis, value = row[11:].split("=")
			folds.append((axis, int(value)))
		return dict(
			coords=coords,
			folds=folds,
		)


def solve(data):
	coords = data['coords']
	folds = data['folds']
	for axis, value in folds:
		print(len(coords))
		newcoords = list()
		for x, y in coords:
			#print(x,y)
			if axis == 'y':
				y = value - abs(y - value)
			else:
				x = value - abs(x - value)
			newcoords.append((x,y))
			
		coords = list(set(newcoords))
	print(len(coords))
	return coords

data = read_input(INFILE)
res = solve(data)

xs, ys = zip(*res)
w = max(xs)+1
h = max(ys)+1
arr = [' '] * w * h
for x, y in res:
	arr[y*w+x] = 'X'

for i in range(h):
	print(''.join(arr[i*w:(i+1)*w]))
